### PR TITLE
Run basic GCP batch job

### PR DESCRIPTION
**Pull request by nweires**
**This change**
- Sets up basic GCP batch job within `buildstock_gcp`. This works, running 3 trivial jobs that just write the output of `openstudio --help` to files in GCS.

- Adds the option to list past jobs, to make it easy to check your job's state.
- Adds `BuildStockBatch / GCP` ascii logo (feel free to improve!)
- Adds `--verbose` flag for optional DEBUG logs, because there were multiple places where I wanted to print short things by default, with the option for the full details when needed.

**Next steps**
I left a lot of TODOs, especially related to these next steps:
- Implement `run_job()` and the sampling steps.
- Replace many hardcoded values with better defaults and/or configurable values. (how many tasks to split the job into, what machine type to use, timeouts, etc)
- Add Terraform config to automatically set up resources that don't exist

**Tested**
Ran locally (minus the docker steps) and confirmed that jobs are succeeding on GCP.
Links for one run:
- [Batch job](https://console.cloud.google.com/batch/jobsDetail/regions/us-central1/jobs/natalie-test23-10-20-153727/details?project=buildstockbatch-dev)
- [Output files](https://console.cloud.google.com/storage/browser/testing-natalie/natalie-test23-10-20-153727;tab=objects?project=buildstockbatch-dev&prefix=&forceOnObjectsSortingFiltering=false)
- Notification messages in Pub/Sub:
image

### CODE REVIEW

File: buildstockbatch/gcp/gcp.py [hunk 1]
```diff
@@ -16,13 +16,17 @@
...
 import argparse
+from datetime import datetime
 import docker
 import json
 import logging
 import os
 import pathlib
 import re
 
+from google.cloud import batch_v1
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py:28]: I'm not sure what the convention is, but maybe this should be within the rest of the imports (rather than its own dedicated block)?

File: buildstockbatch/gcp/gcp.py [hunk 2]
```diff
@@ -16,13 +16,17 @@
...
+
+from buildstockbatch.aws.aws import DockerBatchBase
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: Why is this importing AWS' DockerBatchBase (and how does that work when there's already a DockerBatchBase in this file)?

File: buildstockbatch/schemas/v0.3.yaml [hunk 3]
```diff
@@ -19,11 +19,18 @@ postprocessing: include('postprocessing-spec', required=False)
 references: map(required=False)
 ---
 gcp-spec:
-  job_identifier: regex('^[a-zA-Z]\w{,9}$', required=True)
+  job_identifier: regex('^[a-zA-Z_-]\w{,20}$', required=True)
```
Comment (lathanh) [inline buildstockbatch/schemas/v0.3.yaml]: Intentionally changing this to allow a job_identifier to start with hyphen and underscore? why?

Also, this is actually allowing the job_identifier to be 21 characters.

And what's the motivation of this change? I think we should avoid divergence with the AWS script unless necessary.

File: buildstockbatch/schemas/v0.3.yaml [hunk 4]
```diff
@@ -19,11 +19,18 @@ postprocessing: include('postprocessing-spec', required=False)
...
+  gcp_project: str(required=True)
```
Comment (lathanh) [inline buildstockbatch/schemas/v0.3.yaml]: why is this prefixed with 'gcp_' but the others not? I actually agree the clarity is good, but wonder if we can be more consistent — could it make sense to prefix the other things that are GCP parameters, like region and artifact_registry?

File: buildstockbatch/gcp/gcp.py [hunk 5]
```diff
@@ -53,23 +57,44 @@ def docker_image(self):
...
+        # TODO: how short does this really need to be? What characters are allowed?
+        self.job_identifier = re.sub('[^0-9a-zA-Z-]+', '_', self.cfg['gcp']['job_identifier'])[:20]
 
         self.project_filename = project_filename
+        self.gcp_project = self.cfg['gcp']['gcp_project']
         self.region = self.cfg['gcp']['region']
-        self.project = self.cfg['gcp']['project']
         self.ar_repo = self.cfg['gcp']['artifact_registry']['repository']
+        self.bucket = self.cfg['gcp']['gcs']['bucket']
+        self.prefix = self.cfg['gcp']['gcs']['prefix']
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: imho, I think it'd be clearer to prefix these two properties with 'gcs_'. 🤷

File: buildstockbatch/gcp/gcp.py [hunk 6]
```diff
@@ -216,6 +241,160 @@ def push_image(self):
...
+        )
+
+        # Specify types of VMs to run on
+        # TODO: look into best default machine type for running OpenStudio, but also allow
+        # changing via the project config. https://cloud.google.com/compute/docs/machine-types
+        policy = batch_v1.AllocationPolicy.InstancePolicy(
+            machine_type = 'e2-standard-2',
+        )
+        instances = batch_v1.AllocationPolicy.InstancePolicyOrTemplate(policy=policy)
+        allocation_policy = batch_v1.AllocationPolicy(instances = [instances])
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: May as well set the provisioning_models since you've added 'use_spot' to the schema? Or a TODO at least.

File: buildstockbatch/gcp/gcp.py [hunk 7]
```diff
@@ -249,8 +428,11 @@ def main():
         },
     })
     print(GcpBatch.LOGO)
-    if 'GCP_BATCH_JOB_ARRAY_INDEX' in os.environ:
-        job_id = int(os.environ['GCP_BATCH_JOB_ARRAY_INDEX'])
+    if 'BATCH_TASK_INDEX' in os.environ:
+        # If this var exists, we're inside a single batch task.
+        # TODO: Would it be cleaner to move this to a main() in another file?
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: I propose omitting potential improvements TODOS. There are a _ton_ of possible improvements and getting to them is fairly speculative at this point (then there's the dimension of sharing such improvements with the AWS, Eagle, and local scripts). Meanwhile, they're diluting the TODOs we really need to do.

File: buildstockbatch/gcp/gcp.py [hunk 8]
```diff
@@ -264,12 +446,23 @@ def main():
...
         )
+        parser.add_argument(
+            '--list_jobs',
+            help='List existing jobs',
+            action='store_true'
+        )
+        parser.add_argument(
+            '-v', '--verbose',
+            action='store_true',
+            help='Verbose output'
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: Maybe clarify this makes logging include DEBUG rather than just INFO?

File: buildstockbatch/gcp/gcp.py [hunk 9]
```diff
@@ -216,6 +241,160 @@ def push_image(self):
                     last_status = y['status']
 
 
+    def cleanup(self):
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: can we stick with 'clean' for consistency with the AWS script?

**Review by lathanh — APPROVED** [review]

File: buildstockbatch/gcp/gcp.py (hunk 1 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py:28]: It's separate because it's a third party import. The convention is:
> Imports should be grouped in the following order:
> 1. Standard library imports.
> 2. Related third party imports.
> 3. Local application/library specific imports.
>
> You should put a blank line between each group of imports.

[source](https://peps.python.org/pep-0008/#imports)

File: buildstockbatch/gcp/gcp.py (hunk 2 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: Removed. (Forgot to remove this when I copied it over.)

File: buildstockbatch/gcp/gcp.py (hunk 5 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: Done.

File: buildstockbatch/gcp/gcp.py (hunk 9 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: Done.

File: buildstockbatch/gcp/gcp.py (hunk 6 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: It wasn't immediately obvious how to do this (which is suspicious 🤔), so added a TODO for now.

File: buildstockbatch/gcp/gcp.py (hunk 7 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: That's fair - removed.

File: buildstockbatch/gcp/gcp.py (hunk 8 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: Done.

File: buildstockbatch/schemas/v0.3.yaml (hunk 3 above)
Comment (nweires) [reply buildstockbatch/schemas/v0.3.yaml]: Oops, I confused this with the regex match in gcp.py (which I updated to allow hyphens).

I allowed longer names because it was annoying that my job IDs were being truncated, and I haven't encountered any length limits (yet) in GCP. But I suppose it makes sense to just match AWS, for consistency, so changing it back.

File: buildstockbatch/schemas/v0.3.yaml (hunk 4 above)
Comment (nweires) [reply buildstockbatch/schemas/v0.3.yaml]: I guess the `gcp_` prefix isn't really necessary since these are all nested within the GCP spec, so I'll remove it. I originally added it to help distinguish from the other type of project (i.e. `project_directory` on line 3)

**Review by nweires — COMMENTED** [review]
I also cleaned up the formatting and some unused imports. The only style guidelines I could find for this repo was this config with a 120-character line limit ([here](https://github.com/NREL/buildstockbatch/blob/develop/.flake8)), so I used [Black](https://pypi.org/project/black/) and confirmed that Flake8 didn't give any new errors (based on [this](https://github.com/NREL/buildstockbatch/blob/29958ca29548e25c3222acff68933b55fe15bbca/.github/workflows/ci.yml#L46)).

File: buildstockbatch/gcp/gcp.py (hunk 1 above)
Comment (lathanh) [reply buildstockbatch/gcp/gcp.py:28]: I don't think they've been separating out 3rd-party commits ([their aws.py](https://github.com/NREL/buildstockbatch/blob/develop/buildstockbatch/aws/aws.py), for example)

File: buildstockbatch/schemas/v0.3.yaml [hunk 10]
```diff
@@ -19,14 +19,14 @@ postprocessing: include('postprocessing-spec', required=False)
...
-  gcp_project: str(required=True)
+  job_identifier: regex('^[a-zA-Z_]\w{,9}$', required=True)
+  project: str(required=True)
   region: str(required=True)
   artifact_registry: include('gcp-ar-spec', required=True)
   batch_array_size: num(min=1, max=10000, required=True)
   notifications_email: regex('^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$', name='email', required=True)
   gcs: include('gcs-spec', required=True)
-  use_spot: bool(required=False)
+  use_spot: bool(required=True)
```
Comment (lathanh) [inline buildstockbatch/schemas/v0.3.yaml:29]: Why is it required (and inconsistent with aws-spec)?

File: buildstockbatch/schemas/v0.3.yaml [hunk 11]
```diff
@@ -36,7 +36,7 @@ gcp-ar-spec:
   repository: str(required=True)
 
 aws-spec:
-  job_identifier: regex('^[a-zA-Z]\w{,9}$', required=True)
+  job_identifier: regex('^[a-zA-Z_]\w{,9}$', required=True)
```
Comment (lathanh) [inline buildstockbatch/schemas/v0.3.yaml]: Why adding the ability for the job_identifier to start with an underscore?

File: buildstockbatch/schemas/v0.3.yaml (hunk 3 above)
Comment (lathanh) [reply buildstockbatch/schemas/v0.3.yaml]: Do we need to add the ability for it to start with an underscore? The current AWS regex allows underscores, just not as the first character.

File: buildstockbatch/gcp/gcp.py (hunk 6 above)
Comment (lathanh) [reply buildstockbatch/gcp/gcp.py]: [I think](https://cloud.google.com/python/docs/reference/batch/latest/google.cloud.batch_v1alpha.types.AllocationPolicy.InstancePolicy) you add something like `provisioning_model = ProvisioningModel.SPOT` to the `policy` object at line 344.

But fine to leave as TODO for now!

**Review by lathanh — APPROVED** [review]

File: buildstockbatch/schemas/v0.3.yaml (hunk 11 above)
Comment (nweires) [reply buildstockbatch/schemas/v0.3.yaml]: Fixed.

File: buildstockbatch/schemas/v0.3.yaml (hunk 10 above)
Comment (nweires) [reply buildstockbatch/schemas/v0.3.yaml:29]: The aws-spec is wrong. 🙂  The job fails ([here](https://github.com/NREL/buildstockbatch/blob/661f66abc1e78faf47c9ba56c07ba0918278a15c/buildstockbatch/aws/aws.py#L1696)) if it doesn't exist.

File: buildstockbatch/gcp/gcp.py (hunk 6 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: Yeah, I found that, but that link is for a different version of the library (batch_v1alpha vs batch_v1). In batch_v1, the ProvisioningModel class exists, but there isn't a (documented) way to pass it into the instance policy.

File: buildstockbatch/schemas/v0.3.yaml (hunk 10 above)
Comment (lathanh) [reply buildstockbatch/schemas/v0.3.yaml:29]: (As long as we're diverging) I think it'd be better to leave the config optional and make the code work with it non-present (looks like you can use [`get(key[, default])`](https://docs.python.org/3/library/stdtypes.html#dict.get)).

File: buildstockbatch/gcp/gcp.py (hunk 6 above)
Comment (lathanh) [reply buildstockbatch/gcp/gcp.py]: [non-alpha version looks the same](https://cloud.google.com/python/docs/reference/batch/latest/google.cloud.batch_v1.types.AllocationPolicy.InstancePolicy), no?

Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: Ah, got it. 🤦‍♀️  It's an attribute of InstancePolicy, whereas in the alpha version it's also an attribute of AllocationPolicy, which is where I was looking for it. Will add!
